rcl/src/rcl/parameter_client.c

```diff
@@ -0,0 +1,301 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rmw/qos_profiles.h"
+
+#include <rcl_interfaces/srv/get_parameters.h>
+#include <rcl_interfaces/srv/get_parameter_types.h>
+#include <rcl_interfaces/srv/list_parameters.h>
+#include <rcl_interfaces/srv/set_parameters.h>
+#include <rcl_interfaces/srv/set_parameters_atomically.h>
+
+#include <rcl_interfaces/msg/parameter_type__functions.h>
+
+#include "rcl/error_handling.h"
+#include "rcl/parameter_client.h"
+#include "rcl/subscription.h"
+
+#include "./common.h"
+
+typedef struct rcl_parameter_client_impl_t
+{
+  rcl_node_t * node;
+  rcl_client_t get_client;
+  rcl_client_t get_types_client;  // Also referred to as describe parameters in docs
+  rcl_client_t set_client;
+  rcl_client_t set_atomically_client;
+  rcl_client_t list_client;
+
+  rcl_subscription_t event_subscription;
+
+  // Storage for requests/responses for each client
```

Comment:
Basically I did this so that the parameter client could internally handle filling and accessing the request/response structures so that the API directly accesses the fields that are interesting, and the client library calling rcl doesn't need to instantiate storage for this. But, if I change the take/send calls of the rcl parameter API to be type-erased, then this probably won't be relevant anymore.